src/pymor/algorithms/bfgs.py

```diff
@@ -0,0 +1,320 @@
+    hessian
+        The Hessian matrix.
+    direction
+        The descent direction used to update the parameter.
+    active
+        The active set computed from the |ParameterSpace|.
+    inactive
+        The inactive set computed from the |ParameterSpace|.
+
+    Returns
+    -------
+    hessian
+        The projection of the Hessian applied to the direction.
+    """
+    direction_A = np.multiply(active, direction)
+    direction_I = np.multiply(inactive, direction)
+    hessian = direction_A + np.multiply(inactive, hessian.dot(direction_I))
+    return hessian
+
+
+def update_hessian(hessian, mu, old_mu, gradient, old_gradient):
+    """Update Hessian matrix using BFGS iteration.
+
+    Parameters
+    ----------
+    hessian
+        The current Hessian matrix.
+    mu
+        The current `mu` parameter.
+    old_mu
+        The previous `mu` parameter.
+    gradient
+        The current gradient with respect to the parameter.
+    old_gradient
+        The previous gradient with respect to the parameter.
+
+    Returns
+    -------
+    hessian
+        The next BFGS Hessian matrix iterate.
+    """
+    gradient_difference = gradient - old_gradient
+    mu_difference = mu - old_mu
+    gradient_mu_coefficient = np.dot(gradient_difference, mu_difference)
+
+    if gradient_mu_coefficient > 0.:
+        hessian_grad_diff = np.dot(hessian, gradient_difference)
+        quadratic_coefficient = np.dot(gradient_difference, hessian_grad_diff)
+
+        hessian_part = np.outer(hessian_grad_diff, mu_difference) + np.outer(mu_difference, hessian_grad_diff)
+        mu_part = np.outer(mu_difference, mu_difference)
+
+        mu_coeff = (gradient_mu_coefficient + quadratic_coefficient) / gradient_mu_coefficient**2
+        hessian_coeff = 1. / gradient_mu_coefficient
+
+        hessian += mu_coeff * mu_part - hessian_coeff * hessian_part
+    else:
+        hessian = np.eye(hessian.shape[0])
+
+    return hessian
+
+
+@defaults('miniter', 'maxiter', 'rtol', 'tol_sub', 'stagnation_window', 'stagnation_threshold')
+def error_aware_bfgs(model, parameter_space=None, initial_guess=None, miniter=0, maxiter=100, rtol=1e-16,
+         tol_sub=1e-8, line_search_params=None, stagnation_window=3, stagnation_threshold=np.inf,
+         error_aware=False, error_criterion=None, beta=None, radius=None, return_stages=False):
+    """BFGS algorithm.
+
+    This method solves the optimization problem ::
+
+        min J(mu), mu in C
+
+    for an output functional depending on a box-constrained `mu` using the BFGS method.
+
+    In contrast to `scipy.optimize` with the `L-BFGS-B` methods, this BFGS implementation is
+    explicitly designed to work with an error estimator. In particular, this implementation
+    terminates if the higher level TR boundary from `pymor.algorithms.tr` is reached instead of
+    continuing to optimize close to the boundary.
+
+    Parameters
+    ----------
+    model
+        The |Model| used for the optimization.
```

Comment:
No model is mentioned in the problem definition above. Also, is there really a need to have a `J` here, which comes from a model?

Reply:
The `model` basically serves as a convenient way to:
- provide the output functional $J$,
- compute output derivatives, and
- parse parameter values.

In particular the last two are of interest, with the parameter parsing serving as an assured proxy in case the user did not specify a `ParameterSpace`.

We were thinking about making this function independent of a `model`, however there were some problems with that approach, hence we stuck with the `model`.